client/gutenberg/editor/init.js

```diff
@@ -0,0 +1,76 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import { use, plugins, dispatch } from '@wordpress/data';
+import { once } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import { applyAPIMiddleware } from './api-middleware';
+import { isEnabled } from 'config';
+import debugFactory from 'debug';
+
+const debug = debugFactory( 'calypso:gutenberg' );
+
+// List of Core blocks that can't be enabled on WP.com (e.g for security reasons).
+// We'll have to provide A8C custom versions of these blocks.
+const WPCOM_UNSUPPORTED_CORE_BLOCKS = [
+	'core/file', // see D19851 for more details.
+];
+
+const loadA8CExtensions = () => {
+	require( '../extensions/classic-block/editor' );
+
+	if ( isEnabled( 'gutenberg/block/jetpack-preset' ) ) {
+		require( 'gutenberg/extensions/presets/jetpack/editor.js' );
+	}
+};
+
+// We need to ensure that his function is executed only once to avoid duplicate
+// block registration, API middleware application etc.
+export const initGutenberg = once( ( userId, siteSlug ) => {
```

Comment:
`once` is not applied unique per argument set. The function will be called _once ever_.

How then do we ensure `applyAPIMiddleware( siteSlug )` is called if I edit a post for a second site after I'd already initialized it once before for a first site?

Instead you might want to use a memoization utility capable of handling multiple argument. Lodash's does not. Gutenberg uses (my) [memize](https://github.com/aduth/memize) to this effect.

```
⇒ n_      
n_ > const initGutenberg = ( userId, siteSlug ) => console.log( 'called with', userId, siteSlug );
undefined
n_ > const initGutenbergWithOnce = _.once( initGutenberg );
undefined
n_ > initGutenbergWithOnce( 1, 'example1.wordpress.com' )
called with 1 example1.wordpress.com
undefined
n_ > initGutenbergWithOnce( 1, 'example2.wordpress.com' )
undefined
n_ > const initGutenbergWithLodashMemoize = _.memoize( initGutenberg );
undefined
n_ > initGutenbergWithLodashMemoize( 1, 'example1.wordpress.com' )
called with 1 example1.wordpress.com
undefined
n_ > initGutenbergWithLodashMemoize( 1, 'example2.wordpress.com' )
undefined
n_ > const initGutenbergWithMemize = require( 'memize' )( initGutenberg );
undefined
n_ > initGutenbergWithMemize( 1, 'example1.wordpress.com' )
called with 1 example1.wordpress.com
undefined
n_ > initGutenbergWithMemize( 1, 'example2.wordpress.com' )
called with 1 example2.wordpress.com
undefined
```

Or, if the rest of the code is not meant to / cannot be run, save for the API middleware application, maybe it ought to be pulled out and applied separately.

Reply:
The `initGutenberg` function is a mix of initialization tasks we really want to do just once per app lifetime (like various `require`s and module inits) and other tasks we want to do per site or even per editor instance.

The `siteSlug` here is used to rewrite the URLs of API requests and point them to the per-site WP.com endpoints. `once` means that these request will be made to the first site visited, forever, even after site switch. We can rewrite that to a callback function:
```js
() => getSelectedSiteSlug( store.getState() );
```
Then each request will use the the correct site ID, but will still write the results to the singleton `core/data` store, mixing results from different sites. Not exactly an improvement. Memoizing the init function better won't help either, as long as the data stores remain global singletons.

Reply:
Good catch @aduth! I haven't considered that while refactoring the initialization.

> Or, if the rest of the code is not meant to / cannot be run, save for the API middleware application, maybe it ought to be pulled out and applied separately.

I think it's fine to run the rest only once - for example, we don't want to attempt to register core blocks twice (would generate warnings) etc. So it looks like we should only pull out API middleware application.

Reply:
> Then each request will use the the correct site ID, but will still write the results to the singleton core/data store, mixing results from different sites. Not exactly an improvement. Memoizing the init function better won't help either, as long as the data stores remain global singletons. 

@jsnajdr I'm not sure how to resolve this, given the way core data stores and packages are currently working. Do you think that extending this PR by @Copons to reinitialize the editor on each site switch could work? - https://github.com/Automattic/wp-calypso/pull/29310

Reply:
I think the combination of the following two changes should work:
1. Reset the `core/data` store on each site switch. `core/data` is currently not among the resettable stores in #29310 and needs reset only on site switch, not on each editor load.
2. Instead of passing a constant `siteSlug` to the `wpcomPathMappingMiddleware`, pass a `getSiteSlug` callback that returns the latest `siteSlug` on every call. That makes the middleware map to a different path on a site switch.

If you end up `wpcomPathMappingMiddleware`, I propose the following API change: instead of this:
```js
wpcomPathMappingMiddleware = ( options, next, getSiteSlug ) => ...
use( partialRight( wpcomPathMappingMiddleware, getSiteSlug ) );
```
do this:
```js
wpcomPathMappingMiddleware = getSiteSlug => ( options, next ) => ...
use( wpcomPathMappingMiddleware( getSiteSlug ) );
```
(if you agree it looks like an improvement)

Reply:
> Instead of passing a constant siteSlug to the wpcomPathMappingMiddleware, pass a getSiteSlug callback that returns the latest siteSlug on every call. That makes the middleware map to a different path on a site switch.

Landed this in https://github.com/Automattic/wp-calypso/pull/29368. I'm going to look into point 1 next.

Reply:
The first part is resolved in https://github.com/Automattic/wp-calypso/pull/29445 by @Copons.

Reply:
_should be_ 😄